Add changes to the release process (#19)
php/policies: release-process.rst

.. contents::
   :depth: 2

Sources

+--------------+---------------------------------------------------------------+
| Original RFC | -  `Request for Comments: Release Process                     |
|              |    <https://wiki.php.net/rfc/releaseprocess>`_                |
+--------------+---------------------------------------------------------------+
| Updates      | -  `RFC: Release Cycle Update                                 |
|              |    <https://wiki.php.net/rfc/release_cycle_update>`_          |
|              | -  `RFC: Policy Release Process Update                        |
|              |    <https://wiki.php.net/rfc/policy-release-process-update>`_ |

> +--------------+---------------------------------------------------------------+

Reviewer: Can we not use `list-table`? changes to formatted tables become very noisy.

Reviewer 2: I would just have done this:

```
*Original RFC:* `Request for Comments: Release Process <https://wiki.php.net/rfc/releaseprocess>`_

*Updated By:*

- `RFC: Release Cycle Update <https://wiki.php.net/rfc/release_cycle_update>`_
- `RFC: Policy Release Process Update <https://wiki.php.net/rfc/policy-release-process-update>`_
```

It doesn't have to be complicated.

Author: I tried and it looks ugly.

Author: See https://github.com/php/policies/blob/5a4bdb7a5d518665bf806cdf3f7455025d892ea4/release-process.rst . Basically it looks like it's part of the content.

:Updated by:
   https://wiki.php.net/rfc/release_cycle_update

>    https://wiki.php.net/rfc/policy-release-process-update

Reviewer 2: That's sadly not the right syntax — it puts both things in the same box: https://github.com/php/policies/blob/20dc00d431ee591b2b10221be44fbd75806ec739/release-process.rst#id16

I think duplicating `:Updated by:` would do the trick.

Author: Hmm it shows 2 update by. How about changing it to bullet points instead:

```
**From:**  
`https://wiki.php.net/rfc/releaseprocess <https://wiki.php.net/rfc/releaseprocess>`_

**Updated by:**

- `RFC: Release Cycle Update <https://wiki.php.net/rfc/release_cycle_update>`_
- `RFC: Policy Release Process Update <https://wiki.php.net/rfc/policy-release-process-update>`_
```
?

Reviewer 2: Don't care what the syntax used is, as long as it works. I would probably also bulletise the **From** though.

No feature addition after final x.y.0 release (or x.0.0).

> Backward Compatibility

Reviewer 2: You should also add the new RFC to line 12, under `Updated by`.

======================

A backward compatibility (BC) break is defined as any change that prevents
existing, valid, userland code from continuing to behave as it did in a
previous version within the same major release.

The following are not considered BC breaks:

-  Adding deprecations. Code that triggers a deprecation warning continues to
   work and is still valid. Converting deprecations into exceptions is a user
   choice and not part of the language's default behavior.

> -  Adding new symbols (e.g., functions, classes, constants), even if they may

Reviewer 2: Should that be "new symbols (...) **into the global namespace**, even..." — that is the rule that we have had documented before: https://www.php.net/manual/en/userlandnaming.php

Author: Hmm what about Random namespace used by random extension or newly proposed Uri namespace?

Reviewer 3: That manual section should probably be updated. It's still referring to CODING_STANDARDS rather than policies.

For extensions, the “naming” policy effectively already defines that the extension owns the namespace (https://github.com/php/policies/blob/main/coding-standards-and-naming.rst#namespace-collisions), but I think it is useful to spell it out explicitly once more in the “breaking changes” policy.

Author: So maybe just word it: **into the global or extension namespace**?

version within the same major release.

The following are not considered BC breaks:

-  Adding deprecations. Code that triggers a deprecation warning continues to
   work and is still valid. Converting deprecations into exceptions is a user
   choice and not part of the language's default behavior.

-  Adding new symbols (e.g., functions, classes, constants) into the global or
   core extension namespace, even if they may conflict with user-defined names.
   While these additions can cause name conflicts, they are not classified as BC
   breaks. RFCs and contributors SHOULD make a best effort to minimize the risk
   of conflicts when choosing new names, but SHOULD NOT pick significantly worse
   names purely to reduce conflict risk.

-  Behavior changes in undefined or undocumented edge cases MAY be allowed if
   well justified. However, care SHOULD be taken to minimize disruption.

Major Version Number
====================

-  x.y.z to x+1.0.0

   -  It SHALL included bugfixes and new features.
   -  Extensions support MAY be ended (moved to PECL)
   -  Backward compatibility MAY be broken
   -  API compatibility MAY be broken (internals and userland)

>    -  ABI MAY be broken (internals)

Reviewer 4: Those three lines may be shortened to just `Backward compatibility MAY be broken`. If not, then the last line should be also described as `ABI compatibility MAY be broken (internals)`.

Reviewer 2: I prefer it as it is — more deliberately spelled out.

Author: hmm ok I used single one but will change it to spell each and specifically note that it's a backward compatibility.

Minor Version Number
====================

> -  x.y.z to x.y+1.z

Reviewer 4: ```suggestion
-  x.y.z to x.y+1.0
```

>    -  It SHOULD included bugfixes and new features.

Reviewer 4: ```suggestion
   -  It SHOULD include bugfixes and new features.
```

   -  Extensions support MAY be ended (moved to PECL).

   -  Syntax backward compatibility SHOULD be preserved - every PHP program that

>       compiles SHOULD continue to compile.

Reviewer: Given this is a SHOULD, I would find it useful to spell out when it is ok to violate this guideline. Namely, any breaking change should go through the RFC process. It should ultimately be up to the community to decide what BC breaks are acceptable. (I see this is later clarified).

>    -  Backwards compatibility breaks in minor versions MUST NOT result in silent

Reviewer 4: ```suggestion
   -  Backward compatibility breaks in minor versions MUST NOT result in silent
```

>    -  Backwards compatibility breaks in minor versions MUST NOT result in silent

Reviewer: Nit: Why does this list include whitespace, but the others don't?

Reviewer 2: It shouldn't include whitespace. It's not semantically correct (you're making a list per bullet point here).

      behavioral differences. Instead any breaking change MUST be "obvious" when

>       executing the program.

Reviewer: Can you clarify what this looks like in practice? This makes sense when introducing new error cases, i.e. things that now throw an exception. But it doesn't make much sense to change the behavior of non-error cases while then emitting some notice or warning, if the code isn't otherwise considered problematic. If it _is_ problematic, behavior should not change and a deprecation should be emitted instead, so it can later be removed.

Author: I think the point here is that it should never change for non error cases. For such cases, the deprecation would be used which is not considered as a BC break.

Author: I noted the error bit there. Let me know if it's ok. Maybe it should be SHOULD?

Reviewer 3: @iluuu1994 Given I was the one who suggested this, to give an example: https://wiki.php.net/rfc/static_variable_inheritance would have been an illegal BC break with this policy.

   -  API backward compatibility breaks SHOULD be limited to extensions, or the

>       API of functions within an extension

Reviewer: I personally found the previous "userland" hint useful. I didn't understand it until I looked at the diff.

   -  ABI and internal API compatibility breaks are NOT RECOMMENDED.

   -  Source compatibility SHOULD be kept if possible, while breakages are

>       allowed

Reviewer: This sounds redundant with the two previous bullet points.

   -  It SHOULD included bug fixes and security patches
   -  New features MUST NOT be added.
   -  Extensions support MUST NOT be removed (like move them to PECL)
   -  Backward compatibility MUST be kept (internals and userland)
   -  ABI and internal API compatibility SHOULD be preserved for high severity
      security issues, and MUST be preserved for all other security issues.

It is critical to understand the consequences of breaking BC, APIs or ABIs (only
internals related). It SHOULD NOT be done for the sake of doing it. RFCs
explaining the reasoning behind a breakage and the consequences along with test
cases and patch(es) should help.

If a high severity security fix requires breaking the internal ABI or API, a
proper migration path MUST be provided, and the impact MUST be minimized as much
as possible. This MUST also be accompanied by additional communication during
the release.

All new user-facing features MUST be mentioned in the `UPGRADING
<https://github.com/php/php-src/blob/master/UPGRADING>`_ document. All API and
ABI breaks MUST be mentioned in the `UPGRADING.INTERNALS
<https://github.com/php/php-src/blob/master/UPGRADING.INTERNALS>`_ document.

> See the following links for explanation about API and ABI:

Reviewer 4: There are few terms of compatibility in this document:
- ABI
- API internal
- API userland
- syntax.

Can all of these be briefly explained (for new contributors) in section Backward Compatibility and part starting on line 122 (old 76) may be removed?

-  http://en.wikipedia.org/wiki/Application_binary_interface

Major Version Number
====================

-  x.y.z to x+1.0.0

   -  It SHALL include bugfixes and new features.
   -  Extensions support MAY be ended (moved to PECL).
   -  SAPI support MAY be ended (removed from the php-src repository).
   -  Userland API backward compatibility MAY be broken.
   -  Internal API backward compatibility MAY be broken.
   -  ABI backward compatibility MAY be broken.

Minor Version Number
====================

-  x.y.z to x.y+1.0

   -  It SHOULD include bugfixes and new features.

   -  Extensions support MAY be ended (moved to PECL).

   -  SAPI support is RECOMMENDED to be kept.

   -  Syntax backward compatibility SHOULD be preserved - every PHP program that
      compiles SHOULD continue to compile.

   -  Backward compatibility breaks in minor versions MUST NOT result in silent
      behavioral differences. Instead any breaking change MUST be "obvious" when
      executing the program. It means it SHOULD either throw exception or
      trigger error.

   -  Userland API backward compatibility breaks SHOULD be limited to

>       extensions, or the API of functions within an extension.

Reviewer 3: I'd like to see *something* about “BC breaks must be simple to work-around in a cross-version way”, e.g. what I mentioned in https://externals.io/message/127328#127335.

Author: @TimWolla What do you think about https://github.com/php/policies/pull/19/commits/a6cd5320fa1cf5c8bee404ea8e2e0dc3942aee3b ? I decided to use SHOULD as it should be more recommendation and a hint for voters to decide about it rather than being fully restrictive. It's also more inline with what we have now.

proper migration path MUST be provided, and the impact MUST be minimized as much
as possible. This MUST also be accompanied by additional communication during
the release.

SAPI removal SHOULD generally occur only in a major version. However, in
exceptional cases where a SAPI presents significant maintenance or security
concerns, an exception MAY be made. In all cases, the removal MUST go through
the RFC process.

All new user-facing features MUST be mentioned in the `UPGRADING
<https://github.com/php/php-src/blob/master/UPGRADING>`_ document.

All API and ABI breaks MUST be mentioned in the `UPGRADING.INTERNALS
<https://github.com/php/php-src/blob/master/UPGRADING.INTERNALS>`_ document.

Further reading:

-  http://en.wikipedia.org/wiki/Application_programming_interface
-  http://en.wikipedia.org/wiki/Application_binary_interface

Major Version Number
====================

-  x.y.z to x+1.0.0

   -  It SHALL include bugfixes and new features.
   -  Extensions support MAY be ended (moved to PECL).
   -  SAPI support MAY be ended (removed from the php-src repository).
   -  Userland API backward compatibility MAY be broken.

>    -  Internal API backward compatibility MAY be broken.

Author: Good idea! Added an extra point mentioning the deprecation phase.

>    -  Internal API backward compatibility MAY be broken.

Reviewer: Keeping this slightly vague works for me.

All new user-facing features MUST be mentioned in the `UPGRADING
<https://github.com/php/php-src/blob/master/UPGRADING>`_ document.

All API and ABI breaks MUST be mentioned in the `UPGRADING.INTERNALS
<https://github.com/php/php-src/blob/master/UPGRADING.INTERNALS>`_ document.

Further reading:

-  http://en.wikipedia.org/wiki/Application_programming_interface
-  http://en.wikipedia.org/wiki/Application_binary_interface

Major Version Number
====================

-  x.y.z to x+1.0.0

   -  It SHOULD include bugfixes and new features.
   -  Extensions support MAY be ended (moved to PECL).
   -  SAPI support MAY be ended (removed from the php-src repository).
   -  ABI backward compatibility MAY be broken.
   -  Internal API backward compatibility MAY be broken.
   -  Userland API backward compatibility MAY be broken.
   -  Significant userland API backward compatibility breaks SHOULD be preceded
      by the deprecation phase in the previous major version.

Minor Version Number
====================

-  x.y.z to x.y+1.0

   -  It SHOULD include bugfixes and new features.

   -  Extensions support MAY be ended (moved to PECL).

   -  SAPI support is RECOMMENDED to be kept.

>    -  Internal API compatibility breaks are NOT RECOMMENDED.

Reviewer 5: This does not follow current practice IMHO. We break internal API regularly in minor versions.

Author: Ok I changed it so it's:

```
   -  Internal API compatibility MAY be broken when necessary (for example,
      during refactoring), but SHOULD be preserved whenever possible, such as by
      adding an _ex variant of a function instead of changing the existing one.
```

Does this sound ok to you?

Reviewer 5: This does sound better to me, thank you!

      adding an _ex variant of a function instead of changing the existing one.

   -  ABI backward compatibility MAY be broken.

   -  Syntax backward compatibility SHOULD be maintained - every PHP program
      that compiles SHOULD continue to compile.

   -  Backward compatibility breaks in minor versions SHOULD NOT result in
      silent behavioral differences. Instead any breaking change SHOULD be
      "obvious" when executing the program. It means it SHOULD either throw
      exception or trigger error. Exceptions MAY be considered on a case by case
      basis if the breakage is minimal, the change provides clear user facing
      value, and a non-breaking migration path would result in a significantly
      worse or more complex API.

   -  Backward compatibility breaks introduced in minor versions SHOULD allow
      for straightforward code adjustments to maintain compatibility with both
      the current and upcoming PHP versions. Such adjustments SHOULD be feasible
      with basic tooling and not require complex static analysis.

Patch Version Number
====================

-  x.y.z to x.y.z+1

>    -  It SHOULD include bug fixes and security patches

Reviewer: ```suggestion
   -  It SHOULD include bug fixes and security patches.
```

Backward Compatibility
======================

A backward compatibility (BC) break is any change that causes existing userland
code, which was previously considered valid, to behave differently than before.
This includes changes that result in different output, trigger new errors, or
otherwise alter the observable behavior of the code.

In this context, valid userland code refers to code that compiled and executed
in a previous version without fatal errors. It may or may not have relied on
buggy or unintended behavior. The key consideration is whether the change alters

> the behavior in a way that existing code might have come to depend on.

Reviewer: As discussed in https://github.com/php/php-src/pull/18962, big 👎 for this definition of BC from my side. This disqualifies the majority of bug fixes from being fixes in patches, and potentially forces deprecations and RFCs for things that could previously just be fixed. This helps no one, quite the opposite.

> The key consideration is whether the change alters the behavior in a way that existing code might have come to depend on.

I realize you added this to allow some code fixes even if it changes behavior. But how would you know no code possibly relies on buggy behavior? There's no way to know. This makes this whole clause meaningless.

I would ask what issue we're solving here. Has there been a surge of complaints from bug fixes? I know there have been complaints about the increasing amounts of _intentional_ BC breaks from new major versions. But that's not what we're talking about here.

I will vote against this RFC with this clause.

Author: btw the last change actually relaxed things that were already approved by you so I'm a bit confused why it's suddenly big -1...?

Author: and btw. if you vote against it, it means that you are for disallowing any BC breaks in minor versions as well. That's what the current policy specifies. This PR significantly relaxing this rule and actually allows all those sort of BC breaks happen in the minor version. You should really re-read this before doing such comments.

Author: One thing that I should actually change is the need for RFC for any BC break. That should be a bit more relaxed so we don't need RFC for each such bug fix change.

   -  Bugfixes
   -  New features
   -  Extensions support can be ended (moved to PECL)

>    -  Syntaxt backward compatility MUST be preserved (every PHP program that

Reviewer 6: ```suggestion
   -  Syntax backward compatibility MUST be preserved (every PHP program that
```

Author: Fixed, thanks

   -  Bugfixes
   -  New features
   -  Extensions support can be ended (moved to PECL)
   -  Syntax backward compatibility MUST be preserved (every PHP program that
      compiles must continue to compile)

>    -  API backward compatibility breaks SHOULD be limited to extensions / the

Reviewer 2: Instead of `/ the`, I would use `, or`. (I find it harder to understand what the `/` means here).

Author: Fixed

      API of functions within an extension

>    -  ABI and internal API compatibility may be broken if necessary

Reviewer 2: `compatibility SHOULD NOT be` … RFC 2119 says this for `SHOULD NOT`: 
```
   This phrase, or the phrase "NOT RECOMMENDED" mean that
   there may exist valid reasons in particular circumstances when the
   particular behavior is acceptable or even useful, but the full
   implications should be understood and the case carefully weighed
   before implementing any behavior described with this label.
```

Which is different from `MAY`, which I don't think applies here:
```
   This word, or the adjective "OPTIONAL", mean that an item is
   truly optional.  One vendor may choose to include the item because a
   particular marketplace requires it or because the vendor feels that
   it enhances the product while another vendor may omit the same item.
   An implementation which does not include a particular option MUST be
   prepared to interoperate with another implementation which does
   include the option, though perhaps with reduced functionality. In the
   same vein an implementation which does include a particular option
   MUST be prepared to interoperate with another implementation which
   does not include the option (except, of course, for the feature the
   option provides.)
```

Author: Good point - changed to NOT RECOMMENDED. I also updated README.rst as we haven't linked the  RFC 2119 notation requirements.

   -  Bug fixes and security patches only
   -  Extensions support can't be removed (like move them to PECL)
   -  Backward compatibility must be kept (internals and userland)
   -  ABI and internal API compatibility must be preserved, except in the case

>       of HIGH severity security issues where no other option is possible.

Reviewer 2: RFC 2119 would use `MUST` here, but the except makes it a `SHOULD`. Maybe phrase it as:
```
- ABI and internal API compatibility SHOULD BE preserved for "high" severity security issues, and MUST BE preserved for all other security issues.
```

Author: this sounds better - fixed

It is critical to understand the consequences of breaking BC, APIs or ABIs (only
internals related). It should not be done for the sake of doing it. RFCs
explaining the reasoning behind a breakage and the consequences along with test
cases and patch(es) should help.

> If a HIGH severity security fix requires breaking the internal ABI or API, a

Reviewer 2: I wouldn't capitalise `HIGH` here, as that implies it has a special RFC meaning.

proper migration path must be provided, and the impact should be minimized as
much as possible. This should also be accompanied by additional communication
during the release.

> 

Reviewer 2: - `must be provided` -> `MUST be provided`
- `the impact should be` -> `the inpact MUST be`
- `This should also be` -> `This MUST also be`

`MUST` as used in the RFC is probably more apt, as otherwise people will perhaps just ignore it.

I'm less fuzzy if these paragraphs are considered to be advice, instead of actual rules, but that isn't as clear as it ought to be.

Author: Yeah they are more additional advice but think it make sense to use the key words in them too as they are kind of extra clarifications that should be followed. Changed.

-  x.y.z to x+1.0.0

   -  It SHOULD include bugfixes and new features.
   -  Extensions support MAY be ended (moved to PECL).
   -  SAPI support MAY be ended (removed from the php-src repository).
   -  ABI backward compatibility MAY be broken.
   -  Internal API backward compatibility MAY be broken.
   -  Userland API backward compatibility MAY be broken.

>    -  The significant userland API backward compatibility breaks SHOULD be

Author: Fixed

-  x.y.z to x.y.z+1

   -  It SHOULD include bug fixes and security patches

   -  New features MUST NOT be added.

   -  Extensions support MUST NOT be removed (e.g. moved to PECL).

   -  Userland API backward compatibility MUST be preserved.

>    -  ABI and internal API backward compatibility SHOULD be preserved for fixes

Reviewer: I think this should say reserved rather than preserved?

      -  Release only when there is a security issue or regression issue to
         address.

>       -  Security fix, compatibility build fix and regression fix releases

Reviewer 2: This reads weird without the oxford comma, so I would recommend:

```
Security fix, compatibility build fix, and regression fix releases
```

         SHOULD occur on the same date as bug fix releases for the other
         branches. Exceptions can be made for high risk security issues or high
         profile regressions.

-  Until the end of year 4 (e.g., for PHP 8.4: until Dec 31, 2028):

      -  Security fixes **only**.

      -  Release only when there is a security issue.

>       -  Security fix, compatibility build fix and regression fix releases

Reviewer 2: Same here:
```
Security fix, compatibility build fix, and regression fix releases
```

         SHOULD occur on the same date as bug fix releases for the other
         branches. Exceptions can be made for high risk security issues or high
         profile regressions.

>       -  Regression fixes should be applied only exceptionally for small

Reviewer 2: `should be applied only` -> `SHOULD only be applied`

         regressions or regressions introduced by security fixes and it

>          should get RM approval.

Reviewer 2: Split up the sentence, and the should ought to be a MUST here:
```
regressions or regressions introduced by security fixes.
Regression fixes MUST get RM approval.
```

 Feature selection and development
***********************************

RFCs were introduced many years ago and have proven to be an effective way to
avoid conflicts while providing a structured process for proposing changes to

> php.net. New features or core additions should go through the RFC process. The

Reviewer 3: ```suggestion
the PHP programming language. New features or core additions SHOULD go through the RFC process. The
```
?

Author: Ah yeah this makes more sense. I actually reword it a bit as that SHOULD was clashing slightly with the paragraph below.

 Feature selection and development
***********************************

RFCs were introduced many years ago and have proven to be an effective way to
avoid conflicts while providing a structured process for proposing changes to
the PHP programming language. Most new features or core additions SHOULD go
through the RFC process. However, some features MAY be exempt, as described
below. The process has been used many times for proposing new features and
improvements, even when some proposals were ultimately not accepted.

New features MUST be implemented and proposed using a GitHub pull request.

> Internal API changes (those that do not affect the user-facing API), as well as

Reviewer 2: Do you mean extension-facing APIs as well here?

Author: Yeah I added that explicitly to the patch version rules so it should be covered. I just added a paragraph about updating UPGRADING and UPGRADING.INTERNALS.

Author: I will mark this as resolved but if you think there should be further clarification, please add a new comment / unresolve it.

user-facing features in extensions and SAPIs, do not require an RFC unless a
core developer (someone with commit access to php-src) raises an objection or
requests an RFC within one month of the implementation pull request being

> opened. A core developer MAY also request that the feature be discussed on the

Reviewer 2: How would we know if a PR is opened without being notified? GH produces a lot of notifications and they're easily missed.

I would also start a new paragraph with `A core developer MAY`...

Author: Just to be clear. This is more or less how it currently works. I'm not trying to introduce any changes to the way how it works currently but just reflect it in the policies with some additional clarifications (time limit for the objection and post merge objections). But my intention is not to add some extra burden for developers (e.g. posting about each feature to mailing list).

Personally I think we don't need to notify everyone about every feature. We have code owners for the specific parts and the right people usually get extra notifications (in my case I get emails for parts that I care about but usual GH only notifications for the rest). I also added quite long period for objection which is 1 month to allow extra time for everyone to notice it and still object even after the feature is merged. This should be pretty rare in practice but should allow enough time.

Or do you think that I missed anything that we have in the current process and should be mentioned here?

btw. I split the paragraph as suggested.

-  Until the end of year 3 (e.g., for PHP 8.4: until Dec 31, 2027):

      -  Security fixes, compatibility fixes, fixes to address regressions
         introduced during a normal bug fix release, and fixes for crashes like
         use after free or segmentation faults.

      -  Fixes for crashes SHOULD be applied only exceptionally for small fixes.
         The fix MUST get RM approval.

      -  Updates to ABI incompatible versions of dependent libraries on Windows.

      -  Release only when there is a security issue or regression issue to
         address.

>       -  Security fix, compatibility build fix, and regression fix releases

Author: Makes sense - used compatibility build fix(es) - it's singular in this case because it's a release.
